File: 1604-least-number-of-unique-integers-after-k-removals/least-number-of-unique-integers-after-k-removals.java
import java.util.*;

class Solution {
    public int findLeastNumOfUniqueInts(int[] arr, int k) {
        Map<Integer, Integer> mp = new HashMap<>();
        for (int ele : arr) {
            mp.put(ele, mp.getOrDefault(ele, 0) + 1);
        }
        
        PriorityQueue<Map.Entry<Integer, Integer>> pq = new PriorityQueue<>(
            (a, b) -> Integer.compare(a.getValue(), b.getValue())
        );

        for (Map.Entry<Integer, Integer> entry : mp.entrySet()) {
            pq.offer(entry);
        }

        while (k > 0) {
            Map.Entry<Integer, Integer> entry = pq.poll();
            int freq = entry.getValue();
            if (k >= freq) {
                k -= freq;
            } else {
                pq.offer(entry); // Put back the element as we can't remove all occurrences
                break;
            }
        }

        return pq.size();
    }
}

File: 1604-least-number-of-unique-integers-after-k-removals/least-number-of-unique-integers-after-k-removals.cpp
static int io_opt = []() {
  ios::sync_with_stdio(false);
  return 0;
}();
class Solution {
public:
    int findLeastNumOfUniqueInts(vector<int>& arr, int k) {
       unordered_map<int,int> mp;
       for(int ele:arr){
           mp[ele]++;
       }
       priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;

       for(auto ele:mp){
        //    int freq=ele.second;
        //    int num=ele.first;
           pq.push({ele.second, ele.first});
       }

       while(k){
           int freq=pq.top().first;
           if(k>=freq){
               k=k-freq;
               pq.pop();
           }else break;

       }

       return pq.size();

        
    }
};